src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -129,6 +129,24 @@ private struct Entry {
             }
         }
 
+        public Dictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection):
+            this(collection, null) { }
+
+        public Dictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection, IEqualityComparer<TKey> comparer):
+            this(comparer)
+        {
+            if (collection == null) {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.collection);
+            }
+
+            var coll = collection as ICollection<KeyValuePair<TKey, TValue>>;
+            Initialize(coll?.Count ?? 0);
```

Comment:
I moved cast for ICollection to the place where base constructor is called. If call is successful, I call base ctor with Count value, if not I call base ctor with 0. That way I don't need to check for

`coll !=null and coll.Count > 0`